### ISSUE TITLE
[BUG] gitRepo resources aren't applying to clusters when changing fleet workspace

### ISSUE THREAD

**Issue by slickwarren** [rancher/fleet#1845]
**Rancher Server Setup**
- Rancher version: v2.8-head (fc12e7d)
- Installation option (Docker install/Helm Chart): rke2
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): v1.27.6+rke2r1
- Proxy/Cert Details: valid certs

**Information about the Cluster**
- Kubernetes version: v1.27.6 (rke2, k3s, rke1)
- Cluster Type (Local/Downstream): downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): aws

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: tested with admin

**Describe the bug**

when changing a cluster's workspace, the git-repo resources for fleet aren't applying to the clusters.

**To Reproduce**

* create a cluster (by default, deploys to fleet-default workspace)
* create a new workspace
* in the new workspace, create a git-repo
* transfer the cluster to the new workspace

**Result**
resources from fleet's new workspace aren't deploying. The cluster count remains at 0

**Expected Result**

resources from the gitRepo in the user added workspace should apply to any clusters that are matched. 

**Screenshots**

**Additional context**

adding a git-repo and switching the cluster back to fleet-default results in the expected behavior
tested by using the 'all clusters in workspace' setting and manually selecting a cluster in the workspace, but neither is working

possibly affected by https://github.com/rancher/rancher/issues/43078 as this was found on the same setup

**Comment by slickwarren** [comment]
tested on a released version of rancher, 2.7.8, with a hardened local cluster and am experiencing the same error there:
```
 7 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (containers "working-dir-initializer", "place-tools", "step-git-source", "fleet" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (containers "working-dir-initializer", "place-tools", "step-git-source", "fleet" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "working-dir-initializer", "place-tools", "step-git-source", "fleet" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "working-dir-initializer", "place-tools", "step-git-source", "fleet" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```

I'm actually seeing similar behavior on non-hardened clusters. 
```
Error creating: pods "ctwtest-44c94-nb7c2" is forbidden: violates PodSecurity "restricted:latest": allowPrivilegeEscalation != false (containers "gitcloner-initializer", "fleet" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (containers "gitcloner-initializer", "fleet" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or containers "gitcloner-initializer", "fleet" must set securityContext.runAsNonRoot=true), seccompProfile (pod or containers "gitcloner-initializer", "fleet" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```

**Comment by jiaqiluo** [comment]
here are my 2 cents:
If it is true that Fleet components must run with those mentioned permissions ( allowPrivilegeEscalation, runAsNonRoot, etc.), we have to either modify the PSS level on that namespace or whitelist that namespace.

**Comment by kkaempf** [comment]
@sbulage - can you please verify this bug ?

**Comment by aiyengar2** [comment]
> here are my 2 cents: If it is true that Fleet components must run with those mentioned permissions ( allowPrivilegeEscalation, runAsNonRoot, etc.), we have to either modify the PSS level on that namespace or whitelist that namespace.

The issue here is that these workspaces are user-created resources, so there may be security implications with any automatic process that would attempt to whitelist such a namespace. cc: @macedogm @pjbgf, this may fall in your wheelhouse

**Comment by macedogm** [comment]
We should not add new namespaces to the PSAC exempt list already provided by Rancher, because such namespaces would be user controlled (outside of Rancher's trust boundaries). Additionally, doing such allow lists dynamically is risky.

The recommendation is to fix Fleet's deployment to make sure that all components run with an unprivileged `securityContext`, as Jiaqi and Arvind mentioned.

I haven't dig deep, but if I saw correctly, wouldn't be the case of fixing [GitJob's deployment](https://github.com/rancher/fleet/blob/master/charts/fleet/charts/gitjob/templates/deployment.yaml) to match Fleet's and Fleet-Agent's deployment, respectively?

https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet/templates/deployment.yaml#L62-L68

https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet-agent/templates/deployment.yaml#L29-L35

Note: @aiyengar2 thanks for pinging us on this.

**Comment by raulcabello** [comment]
> We should not add new namespaces to the PSAC exempt list already provided by Rancher, because such namespaces would be user controlled (outside of Rancher's trust boundaries). Additionally, doing such allow lists dynamically is risky.
> 
> The recommendation is to fix Fleet's deployment to make sure that all components run with an unprivileged `securityContext`, as Jiaqi and Arvind mentioned.
> 
> I haven't dig deep, but if I saw correctly, wouldn't be the case of fixing [GitJob's deployment](https://github.com/rancher/fleet/blob/master/charts/fleet/charts/gitjob/templates/deployment.yaml) to match Fleet's and Fleet-Agent's deployment, respectively?
> 
> https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet/templates/deployment.yaml#L62-L68
> 
> https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet-agent/templates/deployment.yaml#L29-L35
> 
> Note: @aiyengar2 thanks for pinging us on this.

This is the `GitJob` deployment, not the k8s `Job` that is created when a `GitRepo` is created or modified. I think we should also add the `securityContext` [here](https://github.com/rancher/fleet/blob/v0.9.0-rc.4/internal/cmd/controller/controllers/git/git.go#L469) and [here](https://github.com/rancher/gitjob/blob/master/pkg/controller/gitjob/gitjobs.go#L294) to fix this issue.

I can't transfer a `Cluster` to a different workspace, but I don't see anything in the logs.

Rancher version: v2.8.0-alpha2
Installation option (Docker install/Helm Chart): rke2  v1.27.6+rke2r1
Downstream cluster: k3s v1.27.6+k3s1

Steps:
- create a cluster (by default, deploys to fleet-default workspace)
- create a new workspace
- in the new workspace, create a GitRepo
- transfer the cluster to the new workspace

Then `Cluster` stays in `Wait Check-In` state and it is never moved to the new workspace I created. However I don't see anything in the gitjob or fleet-controller logs. And the k8s job that should run fleet apply is never created. Where should I look for the error?

**Comment by aiyengar2** [comment]
@raulcabello the relationship between the GitRepo and Job is through the GitJob. When the GitRepo is modified, the related GitJob will also be changed:

https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/internal/cmd/controller/controllers/git/git.go#L433

That’s why the GitJob spawns a new Job.

So it is indeed tied to Job created when a GitRepo is modified.

**Comment by aiyengar2** [comment]
Transferring the Fleet cluster is gated by a feature flag in Rancher, once you enable that in the UI you should be able to execute a transfer.

**Comment by sbulage** [comment]
After enabling the feature flag mentioned in this PR: https://github.com/rancher/dashboard/issues/9730#issuecomment-1749796841and https://github.com/rancher/fleet/issues/1845#issuecomment-1757405438.

It is working as expected. As we can see the cluster can be moved from `fleet-default` workspace to `newly created workspace` (`GitRepo` is already present.).

Cluster details: **It is non-hardened cluster**
- Rancher version: `2.8-head`
- Fleet version: `0.9.0-rc3`
- Upstream cluster RKE2 : `v1.27.6+rke2r1`
- Downstream cluster k3s: `v1.26.4+k3s1`

Also, tried to create new `GitRepo` in the newly created workspace and it is also working as expected.

No `_error or warning_` traces found in `gitjob` pods.

`K8s` job is created and completed with no errors.

We will try with hardened cluster and try to reproduce it.

**Comment by kkaempf** [comment]
It needs to be
* a hardened cluster
* k8s 1.25+
* PSA enforcement needs to be turned on

Also:
> one would not get messages about violates PodSecurity on non-hardened clusters (restricted AdmissionConfiguration
> needs to be set at cluster level, see https://kubernetes.io/docs/tutorials/security/cluster-level-pss/ for more context).

**Comment by raulcabello** [comment]
> We should not add new namespaces to the PSAC exempt list already provided by Rancher, because such namespaces would be user controlled (outside of Rancher's trust boundaries). Additionally, doing such allow lists dynamically is risky.
> 
> The recommendation is to fix Fleet's deployment to make sure that all components run with an unprivileged `securityContext`, as Jiaqi and Arvind mentioned.
> 
> I haven't dig deep, but if I saw correctly, wouldn't be the case of fixing [GitJob's deployment](https://github.com/rancher/fleet/blob/master/charts/fleet/charts/gitjob/templates/deployment.yaml) to match Fleet's and Fleet-Agent's deployment, respectively?
> 
> https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet/templates/deployment.yaml#L62-L68
> 
> https://github.com/rancher/fleet/blob/a1dd929a7f9a8df6ee451dad86174ccd303a60af/charts/fleet-agent/templates/deployment.yaml#L29-L35
> 
> Note: @aiyengar2 thanks for pinging us on this.

@macedogm is it ok if we don't add the `readOnlyRootFilesystem: true`  to the `securityContext`? This is causing problems when cloning git repos with `go-git`. `go-git` is creating temporary files that are causing the problem. `go-git` is a third party library.

The following `securityContext` in both containers in the job created by `gitjob` is working fine in hardened clusters in our test env.
```
									SecurityContext: &corev1.SecurityContext{
										AllowPrivilegeEscalation: &[]bool{false}[0],
										Privileged:               &[]bool{false}[0],
										RunAsNonRoot:             &[]bool{true}[0],
										SeccompProfile: &corev1.SeccompProfile{
											Type: corev1.SeccompProfileTypeRuntimeDefault,
										},
										Capabilities: &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}},
									},
```
Is this `securityContext` enough? see https://github.com/rancher/fleet/pull/1860 and https://github.com/rancher/gitjob/pull/331

**Comment by pjbgf** [comment]
>  This is causing problems when cloning git repos with go-git. go-git is creating temporary files that are causing the problem. go-git is a third party library.

@raulcabello You can mount an `emptyDir` to that specific path (`/tmp`) to bypass this issue.

**Comment by sbulage** [comment]
I have tested with Raul's fix with different images and found that fleet-agent pods are not re-creating with newer images.
Till the time Raul's PR gets merged, he uploaded images and gave it to me for testing.

Throwing below error:
```
Pods "fleet-agent-79fc9f8d57-8k6kw" is forbidden: violates PodSecurity "restricted:latest": unrestricted capabilities (container "fleet-agent" must set securityContext.capabilities.drop=["ALL"]; container "fleet-agent" must not include "ALL" in securityContext.capabilities.add), runAsNonRoot != true (container "fleet-agent" must not set securityContext.runAsNonRoot=false), seccompProfile (pod or container "fleet-agent" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost"):Updated: 0/1
```

I closely look at the PSA exempted namespace [list](https://ranchermanager.docs.rancher.com/assets/files/rancher-psact-22728716f0ef43576c1a3ef2b94fc1b8.yaml) and saw that `cattle-fleet-local-system` namespace which is managed by `Fleet-controller` is absent.
In order test the patch is working, I followed below stesp:
1. I updated PSA file by adding `cattle-fleet-local-system` namespace to exempted namespace list.
2. Restarted `rk2 server`
3. Updated the fleet-agents image
I observed that `fleet-agent` image gets updated without any error.


@pjbgf @macedogm What should I do? should I create new issue for it, if yes, then where? please let me know thanks :smile:

**Comment by sbulage** [comment]
> I have tested with Raul's fix with different images and found that fleet-agent pods are not re-creating with newer images. Till the time Raul's PR gets merged, he uploaded images and gave it to me for testing.
> 
> Throwing below error:
> 
> ```
> Pods "fleet-agent-79fc9f8d57-8k6kw" is forbidden: violates PodSecurity "restricted:latest": unrestricted capabilities (container "fleet-agent" must set securityContext.capabilities.drop=["ALL"]; container "fleet-agent" must not include "ALL" in securityContext.capabilities.add), runAsNonRoot != true (container "fleet-agent" must not set securityContext.runAsNonRoot=false), seccompProfile (pod or container "fleet-agent" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost"):Updated: 0/1
> ```
> 
> I closely look at the PSA exempted namespace [list](https://ranchermanager.docs.rancher.com/assets/files/rancher-psact-22728716f0ef43576c1a3ef2b94fc1b8.yaml) and saw that `cattle-fleet-local-system` namespace which is managed by `Fleet-controller` is absent. In order test the patch is working, I followed below stesp:
> 
> 1. I updated PSA file by adding `cattle-fleet-local-system` namespace to exempted namespace list.
> 2. Restarted `rk2 server`
> 3. Updated the fleet-agents image
>    I observed that `fleet-agent` image gets updated without any error.
> 
> @pjbgf @macedogm What should I do? should I create new issue for it, if yes, then where? please let me know thanks 😄

Ignore this comment as I see `cattle-fleet-local-system` is already exempted. ([Here](https://ranchermanager.docs.rancher.com/reference-guides/rancher-security/psa-restricted-exemptions))

**Comment by slickwarren** [comment]
@manno or @sbulage it doesn't look like rancher has picked up the new RC. Is this something your team can update so that I can run another round of testing?

**Comment by sbulage** [comment]
@slickwarren There was some issue with CI in rancher/charts which seems like it is fixed. It will be available once rancher/charts#3138 merged(hopefully in next hour or so. :crossed_fingers: )

**Comment by sbulage** [comment]
All below test scenario performed on RKE2 cluster with and without hardening it.

**Environment Details:**
- Rancher version: `2.8-head`
- Fleet version: `0.9.0-rc.5`

**QA TEST PLAN**

Scenarios
|Scenario|Test Case|
|---|---|
|1| Test GitRepo deploys application in newly created workspace when cluster is moved from fleet-default workspace to newly created workspace in Hardened cluster.
|2| Test GitRepo deploys application in newly created workspace when cluster is moved from fleet-default workspace to newly created workspace in non-hardened cluster.
|3| Test GitRepo deploys helm application in newly created workspace when cluster is moved from fleet-default workspace to newly created workspace in Hardened cluster.
|4| Test private GitRepo deploys application in newly created workspace when cluster is moved from fleet-default workspace to newly created workspace in Hardened cluster.
|5| Test private GitRepo deploys application in newly created workspace when cluster is moved from fleet-default workspace to newly created workspace non-hardened cluster.

**Comment by kkaempf** [comment]
@slickwarren - please give it a try 😉

**Comment by slickwarren** [comment]
my tests through rancher using rc5 are working well, and I've closed the rancher-side issues. Not sure of fleet's process for closing these, but rancher-qa signs off on this 👍🏼

**Comment by sbulage** [comment]
Thanks @slickwarren I will close this issue now. :+1:
